src/workerd/api/encoding.c++

```diff
@@ -460,38 +464,584 @@ kj::Maybe<jsg::JsString> TextDecoder::decodePtr(
+    //
+    // For Latin-1: ASCII chars (0x00-0x7F) → 1 byte, extended chars (0x80-0xFF) → 2 bytes
+    // Worst-case expansion: 2x, Best-case: 1x (pure ASCII), Typical mixed: ~1.2-1.5x
+    //
+    // Zone 1: "Definitely doesn't fit" (length > bufferSize * 2)
+    //   Even if all ASCII (best case 1:1), string won't fit. Go straight to incremental mode.
+    //   Uses forward scan without length calculation for maximum efficiency.
+    //   Example: 1M chars, 400k buffer → can't possibly fit, scan to find cutoff point
+    //
+    // Zone 2: "Definitely fits" (length * 2 <= bufferSize)
+    //   Even if all extended Latin-1 (worst case 1:2), string will fit. Convert directly.
+    //   Example: 100k chars, 250k buffer → worst case 200k bytes, guaranteed to fit
+    //
+    // Zone 3: "Maybe fits" (bufferSize < length * 2 AND length <= bufferSize * 2)
+    //   Might fit depending on ASCII/extended ratio. Use forward scan with length calculation.
+    //   Avoids redundant work: scanning once gets us both position and UTF-8 length.
+    //   Example: 600k chars, 700k buffer → fits if mostly ASCII, doesn't if mixed
+    //
+    // Threshold selection (bufferSize * 2):
+    //   - Chosen based on worst-case Latin-1 expansion of 2x
+    //   - Optimized for common case: small buffer relative to input (SSR, streaming)
+    //   - Trade-off: Zone 3 still does forward scan, but with length calculation overhead
+    //   - Performance cliff exists for borderline cases (e.g., 1M chars, 500k buffer falls
+    //     into Zone 3), but forward scan with length is still reasonably efficient
+    //
+    // Future optimization: Could use sampling to estimate ASCII ratio and choose zone
+    // dynamically, but adds complexity for marginal benefit in typical workloads.
+
+    if (length > bufferSize * 2) {
+      // Zone 1: Incremental mode - forward scan to find what fits, then convert
+      size_t read = findBestFitLatin1(data, length, bufferSize);
+      size_t written = simdutf::convert_latin1_to_utf8(data, read, outputBuf.begin());
+      return TextEncoder::EncodeIntoResult{
+        .read = static_cast<int>(read),
+        .written = static_cast<int>(written),
+      };
+    }
+
+    if (length * 2 <= bufferSize) {
+      // Zone 2: Fast path - worst-case (2x) definitely fits, convert directly
+      size_t written = simdutf::convert_latin1_to_utf8(data, length, outputBuf.begin());
+      return TextEncoder::EncodeIntoResult{
+        .read = static_cast<int>(length),
+        .written = static_cast<int>(written),
+      };
+    }
+
+    // Zone 3: "Maybe fits" - use forward scan with length calculation to avoid double-scan
+    auto [read, utf8Length] = findBestFitLatin1<true>(data, length, bufferSize);
+
+    // Check if everything fit
+    if (read == length) {
+      // ASCII fast path: utf8Length == length means all chars are ASCII, no conversion needed
+      if (utf8Length == length) {
+        memcpy(outputBuf.begin(), data, length);
```

Comment:
Nit: using `kj::ArrayPtr`'s copyFrom is generally safer/preferred. At the very least this probably needs a `KJ_DASSERT` for a bounds check.